reframe/utility/versioning.py

```diff
@@ -0,0 +1,104 @@
+import abc
+import re
+
+
+class CustomBooleanOperations(abc.ABC):
+    """Abstract class to customize Boolean operations."""
+
+    def __lt__(self, other):
+        return self._compare(other, lambda s, o: s < o)
+
+    def __le__(self, other):
+        return self._compare(other, lambda s, o: s <= o)
+
+    def __eq__(self, other):
+        return self._compare(other, lambda s, o: s == o)
+
+    def __ge__(self, other):
+        return self._compare(other, lambda s, o: s >= o)
+
+    def __gt__(self, other):
+        return self._compare(other, lambda s, o: s > o)
+
+    def __ne__(self, other):
+        return self._compare(other, lambda s, o: s != o)
+
+    @abc.abstractmethod
+    def _compare(self, other, operation):
+        """Define the method for comparing versions.
+
+        This method returns a bool.
+        """
+
+
+class ReleaseTag(CustomBooleanOperations):
+    def __repr__(self):
+        return "release"
+
+    def _compare(self, other, operation):
+        if isinstance(other, self.__class__):
+            return operation(0, 0)
+        elif isinstance(other, DevelopmentTag):
+            return operation(1, 0)
+        else:
+            return NotImplemented
```

Comment:
I don't see the reason for all this infrastructure here. I don't think you need it. The `Version` class itself should implement the rich comparison operators. Also there is no need to define all of them. Python defines `__ne__`  by call inverting the result of the `__eq__` function. For the rest of the comparison operators, you should better use [functools.total_ordering](https://docs.python.org/3.6/library/functools.html#functools.total_ordering).